### PR TITLE
feature(unlock-js): adding support for cardPurchaser in unlock-js

### PR DESCRIPTION
**Pull request by julien51**
# Description

This is the unlockjs part of the universal card support!

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: packages/unlock-js/src/erc20.ts [hunk 1]
```diff
@@ -93,6 +93,96 @@ export async function approveTransfer(
...
+interface TransferAuthorizationMessage {
+  from: string
+  to: string
+  value: any
+  validAfter: number
+  validBefore: number
+  nonce: string
+}
+
+export async function signTransferAuthorization(
```
Comment (julien51) [inline packages/unlock-js/src/erc20.ts:105]: I am not sure this is supported by many ERC20 but I was not sure what the best approach was to keep things simple.

File: packages/unlock-js/src/erc20abi.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
 const erc20abi = [
+  'function name() public view returns (string)',
```
Comment (julien51) [inline packages/unlock-js/src/erc20abi.ts:2]: These are defined for USDC but probably not for _all_ ERC20.

File: packages/unlock-js/src/CardPurchaser.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    const { chainId } = await provider.getNetwork()
+    const contract = new ethers.Contract(
+      cardPurchaserAddress,
+      CardPurchaserAbi,
+      provider
+    )
+
+    const domain = {
+      name: await contract.name(),
+      version: await contract.version(),
```
Comment (searchableguy) [inline packages/unlock-js/src/CardPurchaser.ts:84]: Promise.all?

File: packages/unlock-js/src/erc20.ts [hunk 4]
```diff
@@ -93,6 +93,96 @@ export async function approveTransfer(
...
+  provider: ethers.providers.Provider,
+  signer: ethers.Signer
+) {
+  const { chainId } = await provider.getNetwork()
+  const contract = new ethers.Contract(erc20ContractAddress, erc20abi, provider)
+
+  const domain = {
+    name: await contract.name(),
+    version: await contract.version(),
+    chainId,
```
Comment (searchableguy) [inline packages/unlock-js/src/erc20.ts:117]: promise all here too to speed up.

File: packages/unlock-js/src/erc20.ts [hunk 5]
```diff
@@ -93,6 +93,96 @@ export async function approveTransfer(
...
+  message: TransferAuthorizationMessage,
+  chainId: number,
+  signature: string,
+  provider: any
+) {
+  const contract = new ethers.Contract(erc20ContractAddress, erc20abi, provider)
+
+  const domain = {
+    name: await contract.name(),
+    version: await contract.version(),
```
Comment (searchableguy) [inline packages/unlock-js/src/erc20.ts:146]: same

**Comment by julien51** [comment]
PTAL @clemsos @kalidiagne @searchableguy  :)

File: packages/unlock-js/src/erc20abi.ts (hunk 2 above)
Comment (julien51) [reply packages/unlock-js/src/erc20abi.ts:2]: They are not even in the ERC20 spec...

File: packages/unlock-js/src/erc20.ts [hunk 6]
```diff
@@ -93,6 +93,106 @@ export async function approveTransfer(
...
+  validAfter: number
+  validBefore: number
+  nonce: string
+}
+
+export async function signTransferAuthorization(
+  erc20ContractAddress: string,
+  message: TransferAuthorizationMessage,
+  provider: ethers.providers.Provider,
+  signer: ethers.Signer
```
Comment (searchableguy) [inline packages/unlock-js/src/erc20.ts:109]: Same here. Can we use an object so adding new arguments does not cause a breaking change>

File: packages/unlock-js/src/walletService.ts [hunk 7]
```diff
@@ -1021,4 +1035,71 @@ export default class WalletService extends UnlockService {
...
+   * This yields the authorizations required to spend USDC on behalf of the user
+   * to buy a key from a specific lock. This is mostly used for the universal
+   * card support!
+   * @param param0
+   * @returns
+   */
+  async getAndSignAuthorizationsForTransferAndPurchase({
+    amount,
+    lockAddress,
+    network,
```
Comment (searchableguy) [inline packages/unlock-js/src/walletService.ts:1049]: Can you make it consistent with what we do with other methods?

**Review by searchableguy — APPROVED** [review]
A few consistency changes and question but otherwise LGTM

File: packages/unlock-js/src/CardPurchaser.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+   * These will be submitted to the purchase function on the CardPurchaser contract
+   * @param network
+   * @param lockAddress
+   * @param signer
+   * @returns
+   */
+  async getPurchaseAuthorizationSignature(
+    network: number,
+    lockAddress: string,
+    signer: Signer
```
[the remark this answers is not in the document]
Comment (julien51) [reply packages/unlock-js/src/CardPurchaser.ts:65]: ok ok

**Pull request merged**